### PR TITLE
fix: make EventuallyWithT concurrency safe

### PR DESCRIPTION
**Pull request by czeslavo**
## Summary

In `assert.EventuallyWithT`, `assert.CollectT` is used in a concurrent context which may lead to races as explained in https://github.com/stretchr/testify/issues/1391.

## Changes

We create a `CollectT` per every condition tick and send its errors over a channel. On the reader side of the channel, we cache the `lastTickErrs` so they can be copied over to t once the timeout is fired. The only access to `lastTickErrs` happens in the reader case and the timeout case which both run in the same goroutine.

## Motivation

We cannot use `EventuallyWithT` family of functions/methods in our codebase due to data races it causes.

## Related issues
Closes https://github.com/stretchr/testify/issues/1391.

### CODE REVIEW

**Review by brackendawson — APPROVED** [review]
I was thinking whether this can be fixed by enforcing that condition funcs can never overlap but the documentation is clear that it is checked every tick, so that would be a breaking change. The test relies on the race detector, I don't see a better way to do it.

**Review by brackendawson — APPROVED** [review]

File: assert/assertions.go
Comment (dolmen) [inline assert/assertions.go:1901]: `HasErrors` should be private (`Reset` and `Copy` should have been made private too unfortunately, but they are now part of the public API).

**Review by dolmen — CHANGES_REQUESTED** [review]

**Comment by dolmen** [comment]
So while this effectively serializes calls to CollectT methods (#1418), I'm not convinced this really fixes the core issue.

This implementation doesn't fix the real problem which that we may have multiple instances of the `condition` function running at the same time (overlapping) but we only use a single `CollectT` instance to collect the results, with calls to `Reset` which could alter the result of conditions.

I think we will merge this (with `HasErrors` made private), but work on `EventuallyWithT` must not be considered done.

**Comment by czeslavo** [comment]
> So while this effectively serializes calls to CollectT methods (#1418), I'm not convinced this really fixes the core issue.
> 
> This implementation doesn't fix the real problem which that we may have multiple instances of the `condition` function running at the same time (overlapping) but we only use a single `CollectT` instance to collect the results, with calls to `Reset` which could alter the result of conditions.
> 
> I think we will merge this (with `HasErrors` made private), but work on `EventuallyWithT` must not be considered done.

@dolmen Thanks for the comment, you're right! I've taken a look into that again and came up with a solution that doesn't involve a mutex. It also doesn't allow a situation where we end up with an empty `CollectT` on timeout in case of timeout firing up just after `collect.Reset()` call. We create a `CollectT` per every condition tick and send its errors over a channel. On the reader side of the channel, we cache the `lastTickErrs` so they can be copied over to `t` once the timeout is fired. The only access to `lastTickErrs` happens in the reader case and the timeout case which both run in the same goroutine.

File: assert/assertions_test.go
Comment (dolmen) [inline assert/assertions_test.go:2790]: Don't use `CollectT` here.

**Review by dolmen — CHANGES_REQUESTED** [review]
This looks very good!

This is the solution I want!

Also it will allow us to get rid of `CollectT` which was a design error of the original implementation (a separate work for which I am studying plans since a week).

**Comment by dolmen** [comment]
As the `Reset` and `Copy` methods of `CollectT` are not needed anymore, and they should have been made private in the first place, please replace their implementation with a panic and add a `// Deprecated:` comment.

**Comment by dolmen** [comment]
@czeslavo Note for the future, that completely replacing your original implementation on which at least 2 other people worked too (reviewers) with a completely different one, is not a good practice. It would have been better to open a separate PR and close this one.

Anyway, in this case, the original implementation can be forgotten.

File: assert/assertions_test.go
Comment (czeslavo) [reply assert/assertions_test.go]: Done. Please note this prevents us from checking how many errors were reported in this test. Nevertheless we have others that do check that and this one was meant for triggering the race itself so I guess it can be like that.

Comment (dolmen) [reply assert/assertions_test.go]: I didn't say to not capture errors.

Just don't use CollectT for that.

Because my aim is to deprecate CollectT.

**Review by dolmen — CHANGES_REQUESTED** [review]

**Comment by dolmen** [comment]
"address review comments" is never a good commit message. It is completely useless when looking at the Git log or with "git blame". A good commit message must give a summary of what changes, and, if necessary, why. A commit that ONLY says why is never good.

In the present case squashing the commits is the best thing to do.

File: assert/assertions_test.go
Comment (cszczepaniak) [inline assert/assertions_test.go:2789]: It may be beneficial to have an additional test which explicitly exposes the bug this PR fixes. [Here](https://github.com/stretchr/testify/pull/1438/files#diff-7190065248695df477329c4cada2d71f323ab7940ecffff2d3d5fa43caf79427R2766) is an example -- if the first check of the condition is allowed to complete quickly and the next takes longer than the timeout, it is very easy to expose the bug.

File: assert/assertions.go
Comment (czeslavo) [reply assert/assertions.go]: Addressed.

File: assert/assertions_test.go
Comment (czeslavo) [reply assert/assertions_test.go:2789]: Thank you for the suggestion. Yes, this reproduces the issue of reporting no errors due to the `Reset()` call just before running another condition. Allowed myself to copy the test and supplement that with comments. 👍

File: assert/assertions_test.go
Comment (czeslavo) [reply assert/assertions_test.go]: 👍 Added `errorsCapturingT` mock implementation so that it fits the `EventuallyWithT` needs.

File: assert/assertions_test.go
Comment (dolmen) [inline assert/assertions_test.go:2817]: To better protect against race condition (the implementation of EventuallyWithT doesn't run overlapping conditions for now, but we don't want the test to be sensible to such a change), it would be better to avoid using a variable shared by multiple runs of the condition.

So it would be wise to use a channel of bools instead of `calledOnce`.

```go
mustSleep := make(chan bool, 2)
mustSleep <- false
mustSleep <- true
close(mustSleep)

...

if <-mustSleep {
    time.Sleep(time.Second)
}
```

File: assert/assertions_test.go
Comment (dolmen) [inline assert/assertions_test.go:2820]: Use `Fail` instead. Or directly `collect.Errorf`.

```suggestion
		Fail(collect, "condition fixed failure")
```

File: assert/assertions_test.go
Comment (dolmen) [inline assert/assertions_test.go:2824]: Use your `errorsCapturingT`.

**Review by dolmen — CHANGES_REQUESTED** [review]

**Comment by czeslavo** [comment]
@dolmen Is there anything else that I could do that's blocking this PR from merging?

**Review by dolmen — CHANGES_REQUESTED** [review]

File: assert/assertions.go
Comment (marshall-lee) [reply assert/assertions.go:1934]: Yes this way is more safer and is mandatory if `FailNow()` will use `runtime.Goexit()` as in #1481.

Comment (czeslavo) [reply assert/assertions.go]: Yeah, that makes sense. I've applied the suggestion. 

I'll _defer_ 🙂  implementing the test case ensuring errors are collected when the panic occurs to https://github.com/stretchr/testify/issues/1457 because that will depend on how we will handle it. As I understand, `runtime.Goexit()` in `CollectT.FailNow` will unblock using `require` assertions, but not ANY panic that could occur in conditions. To handle any panic, we could catch the panic in the goroutine running the condition, i.e.:

```go
collect := new(CollectT)
defer func() {
	if err := recover(); err != nil {
		collect.Errorf("Panic caught from condition: %v", err)
	}
	ch <- collect.errors
}()
condition(collect)
```

Comment (dolmen) [reply assert/assertions.go]: @czeslavo You're right. I'll merge as is. I also notice that @marshall-lee has already implemented the defer in #1481 which is in the pipe.

**Review by dolmen — APPROVED** [review]

**Comment by dolmen** [comment]
Thank you @czeslavo and all reviewers!

**Pull request merged**
